Sources/PackageModel/Package.swift

```diff
@@ -129,3 +137,42 @@ extension Package.Error: CustomStringConvertible {
         }
     }
 }
+
+extension DiagnosticsMetadata {
+    public static func packageMetadata(identity: PackageIdentity, location: String) -> DiagnosticsMetadata {
+        var metadata = DiagnosticsMetadata()
+        metadata.packageIdentity = identity
+        metadata.packageLocation = location
+        return metadata
+    }
+}
+
+extension DiagnosticsMetadata {
+    public var packageIdentity: PackageIdentity? {
+        get {
+            self[PackageIdentityKey.self]
+        }
+        set {
+            self[PackageIdentityKey.self] = newValue
+        }
+    }
+    
+    enum PackageIdentityKey: DiagnosticsMetadataKey {
+        typealias Value = PackageIdentity
+    }
+}
+
+extension DiagnosticsMetadata {
+    public var packageLocation: String? {
+        get {
+            self[PackageLocationKey.self]
+        }
+        set {
+            self[PackageLocationKey.self] = newValue
+        }
+    }
+    
+    enum PackageLocationKey: DiagnosticsMetadataKey {
```

Comment:
This PR itself is looking good, with the approach of the typed metadata.  I think a large part of the effort going forward will be to define what the right set of keys are and how they should be used, so that IDEs and other clients know what to look for with predictable semantics (unlike the package location today, which is wildly diverse).  As discussed the metadata can then be used to store both the context and the more immediate location at which the problem was identified.  Some of these will also need to be sets or arrays, e.g. if the error is that there are multiple packages with the same name, for example, then there are multiple relevant locations associated with that (just as a redeclared type or variable in Swift or C).